### PR TITLE
Automated unified CSV generation 

### PR DESCRIPTION
**Pull request by davidvossel**
This PR adds a build flow for automating the generation the ocs-operator's unified CSV.

## Overview
For now, this automation involves two steps. The "two step" requirement is because the PRs that introduce the CSV/CRD manifest templates into the component-level operator container images have not made it into an official release of those operators yet.  

As a result of this, in order to demonstrate the automation I have to use unofficial rook/noobaa container images to source the manifests, and then use the official container images when generating the unified CSV from those container images.

Step 1: **sourcing the component level manifests**
This step extracts the CSV and CRD data from the component-level operator's container images and places them in the temporary directory `_output/csv-templates/*`

Step2: **Generating the unified CSV**
This step takes all the manifests sourced into `_output/csv-templates/*` can merges them into a single unified CSV. The result is stored in the `deploy/olm-catalog/ocs-operator/$CSV_VERSION/` directory just like the ocs-operator's registry build step expects. 

## Example: Creating a release of OCS Operator 0.0.2

Step1: Pick the ROOK and NOOBAA container images to source the CSV/CRD data from.
```
$ export ROOK_IMAGE=quay.io/dvossel/ceph-amd64:latest
$ export NOOBAA_IMAGE=quay.io/dvossel/noobaa-operator:latest
$ make source-manifests
```

Step2: Choose the CSV version and container images to use in the finalized unified CSV. **NOTE: the 'master' and 'latest' floating tags should never be used in an official release.**
```
$ export CSV_VERSION=0.0.2
$ export ROOK_IMAGE=rook/ceph:master
$ export CEPH_IMAGE=ceph/ceph:v14.2.2-20190722
$ export NOOBAA_IMAGE=noobaa/noobaa-operator:1.0.2
$ export OCS_IMAGE=quay.io/ocs-dev/ocs-operator:latest
make gen-csv
```

This example results in both a unified CSV along with all the corresponding CRDs being placed in `deploy/olm-catalog/ocs-operator/0.0.2/` for release.

### CODE REVIEW

**Comment by davidvossel** [comment]
/cc @kshlm @jarrpa 

I'm working on testing the CSV that's generated by this process now. The CSV sourcing/merging logic is ready for review though. 

One thing I noticed is that the Ceph Filesystem CRD isn't being sourced automatically yet. I can fix that, but when I look at how the rook CSV is built, the filesystem CRD isn't included.

We do need that  API in the ocs operator, correct?

**Comment by kshlm** [comment]
Yup. The CephFS CSI driver needs the the CRD. The Rook CSV is not built for CSI drivers yet. OCS only supports CSI, so I've had to make additions in our CSV.

In addition to the CephFS CRD, I also include the RBAC needed for the CSI drivers in our CSV. Also, the rook-ceph deployment is configured differently compared the rook-ceph CSV.

**Comment by davidvossel** [comment]
> In addition to the CephFS CRD, I also include the RBAC needed for the CSI drivers in our CSV. Also, the rook-ceph deployment is configured differently compared the rook-ceph CSV.

ah yes, I see. The ENV VARs the ocs-operator needs for the ceph deployment are custom. I can fix that and make it consistent with what we have today in the ocs 0.0.1 CSV. I'll add the cephFS CRD/RBAC as well. 

One thing I noticed is that the ENV VAR ```ROOK_CSI_ENABLE_CEPHFS``` isn't set in current ocs-operator 0.0.1 CSV. Do we need that variable enabled?

**Comment by kshlm** [comment]
> One thing I noticed is that the ENV VAR `ROOK_CSI_ENABLE_CEPHFS` isn't set in current ocs-operator 0.0.1 CSV. Do we need that variable enabled?

It is supposed to be enabled by default now. But I'm not sure if setting this ENV VAR in the example operator manifest is how this is enabled by default. This may be the reason I see issues getting the CephFS driver to run right now. So there is no harm explicitly enabling `ROOK_CSI_ENABLE_{RBD,CEPHFS}`.

**Comment by kshlm** [comment]
@davidvossel Am I right in assuming that all available manifests for CRDs, CRs, RBAC are bundled in the rook image and just those referred to in the CSV.

**Comment by davidvossel** [comment]
>  Am I right in assuming that all available manifests for CRDs, CRs, RBAC are bundled in the rook image and not just those referred to in the CSV. Just want to know if you need to do any further work on rook to get CephFS included in our CSV.

Unfortunately, I have to go back and work on rook. The current solution sources everything that is bundled in the rook CSV, which doesn't include the cephfs CRD and corresponding RBAC. 

The good news is that once I address this in rook, these scripts automatically pick up the changes. So, there's only one place to make this change. 

For the ceph deployment ENV VAR's, I'm going to make a custom ocs override struct in the csv-merger tool so we have completely control over what features are enabled and how.

**Comment by davidvossel** [comment]
I created https://github.com/rook/rook/pull/3673 to address the missing csi related RBAC roles. We'll be able to source those rules once that rook PR is finalized.

**Comment by davidvossel** [comment]
@jarrpa @kshlm 

I believe this is in fairly good shape now. I need some help verifying all of this works though.

To install using the autogenerated csv, use the `deploy/deploy-with-olm.yaml` script I updated in this branch.

All I have been able to do is ensure that all the operators are installed when I install the new ocs-registry build from the autogenerated CSV. It's unclear what functionality I can verify beyond this. See the results below.

```
$ oc get pods -n openshift-storage
NAME                                  READY   STATUS    RESTARTS   AGE
noobaa-operator-7f5fd5d4f6-kwd8d      1/1     Running   0          20s
ocs-operator-589bb994b8-xcrmm         1/1     Running   0          20s
rook-ceph-operator-67cfcfc467-2rp9x   1/1     Running   0          19s
rook-discover-8wzlb                   1/1     Running   0          18s
rook-discover-ttvcz                   1/1     Running   0          18s
rook-discover-vqg55                   1/1     Running   0          18s
```

**Comment by jarrpa** [comment]
I'll give this a go myself, as well. My plan would be to bring it all up, then try creating a [BlockPool](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/pool.yaml) and try to follow the Ceph-CSI RBD example: https://github.com/rook/rook/tree/master/cluster/examples/kubernetes/ceph/csi/rbd

**Comment by davidvossel** [comment]
> I'll give this a go myself, as well. My plan would be to bring it all up, then try creating a BlockPool and try to follow the Ceph-CSI RBD

Ah, perfect. Thanks for the info.

I just noticed something looking at the PR that still needs to be addressed. We're setting custom ENV VARs on the rook deployment in our manually created CSV. I need to go in and make an ENV VAR override for the rook deployment on the auto-generated CSV for the exact variables we're toggling. 

This PR is a bit of a moving target for me as updates keep coming in to the existing manually created CSV. Tricky stuff :) 

I'll make that rook deployment ENV VAR update first thing next week and give the blockpool a test.

**Comment by kshlm** [comment]
@davidvossel The issues you're seeing with testing this is down to #51 and a big change in Rook recently (over the weekend). #51 adds placement configuration for rook, which is setting an invalid label as a toleration. This is preventing Rook from scheduling the csi pods. Which in turn means the csi configuration isn't available for rook to begin scheduling the other pods. An older version, rook/ceph:v1.0.0-457.gddfb403, allows the csi pods to start, except the rbd-provisioner. It rbd-provisioner-sa has no rules defined. It needs to have the same rules defined as the cephfs-provisioner-sa, with addition of rules needed for snapshots. Not sure why it is empty in the generated csv.

File: deploy/olm-catalog/ocs-operator/0.0.1/ocs-operator.v0.0.1.clusterserviceversion.yaml [hunk 1]
```diff
@@ -594,329 +532,597 @@ spec:
...
+                      fieldPath: metadata.namespace
+                image: noobaa/noobaa-operator:1.0.2
+                imagePullPolicy: IfNotPresent
+                name: noobaa-operator
+                resources:
+                  limits:
+                    cpu: 250m
+                    memory: 256Mi
+              serviceAccountName: noobaa-operator
+      - name: ocs-operator
```
Comment (kshlm) [inline deploy/olm-catalog/ocs-operator/0.0.1/ocs-operator.v0.0.1.clusterserviceversion.yaml]: The deployment of ocs-operator has been added a second time here. Didn't notice it until I tried to set an alternate image manually, and had two ocs-operator pods running.

**Comment by kshlm** [comment]
@davidvossel We would also need to allow setting of ROOK_CSI_CEPH_IMAGE, ROOK_CSI_REGISTRAR_IMAGE, ROOK_CSI_PROVISIONER_IMAGE, ROOK_CSI_SNAPSHOTTER_IMAGE and ROOK_CSI_ATTACHER_IMAGE for the rook-ceph deployment. We need this again to pin the versions of images being used.

**Pull request merged**